### PR TITLE
Body ECS/Refactor

### PR DESCRIPTION
**Pull request by DoctorBeard**
## About the PR 
ECS'd and refactored the body system. The whole thing is based on #6147, though since this was supposed to be a learning experience I reimplemented the changes from scratch rather than try to merge master into that branch. Once I had reimplemented those changes I did some additional refactoring on top of it.

Opening as a draft PR for now because there is some stuff I'm still not super happy with and probably some stuff I did wrong but just don't know how to do better yet in the context of SS14. I also still want to test this a bunch more to make sure I didn't break anything.

Aside from ECSing the thing one of the biggest changes is that body parts live in ContainerSlots now, similar to the ItemSlotsSystem or the inventory slots on mobs. The containers are wrapped in a BodyPartSlot object that tracks some body part slot meta data, like part type and connections. My line of thinking behind that was that body parts were already more or less arranged in slots anyway and rather than having the body system be a homebrew ContainerManager it might as well let the actual ContainerManager do the heavy lifting.

There are two downsides to this approach though and I'm not happy about how I ended up solving them. Feedback and ideas would be much appreciated there. With body parts creating ContainerSlots on the mob entity now there were some conflicts with the existing inventory slots. 

One is the head slot, with both the inventory ContainerSlot (for headgear) and the body part slot being called "head" and trying to create their container with this ID. As a workaround I have changed the ID of the body part to "head-part" but this feels iffy and isn't consistent with the rest of the parts now.

The second one was with the HandsSystem, which creates ContainerSlots for held items for each added hand body part using the same ID as that body part. As a workaround I've made it use "{slotId}-inv" instead but that also feels hacky.

The BodyScanner actually works now but I just gave it the bare minimum amount of work to not be a hollow shell of commented out lines and broken code. It just lists body parts and mechanisms now with no real additional information, and the UI just shows the body parts of whoever last interacted with the device.

Since this is my first big PR and the primary purpose of this refactor for me was to learn and get into the SS14 codebase I'm absolutely open to any feedback and suggestions.

### CODE REVIEW

**Comment by Just-a-Unity-Dev** [comment]
holy shit

**Comment by DoctorBeard** [comment]
Tests are failing because of #10273 by the way. Linter is failing because I'm a dummy and forgot to change something.

File: Content.Shared/Body/Systems/Part/SharedBodyPartSystem.cs [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+            component.PartType = state.PartType;
+            component.Symmetry = state.Symmetry;
+        }
+    }
+
+    protected virtual void OnAddedToBody(EntityUid uid, SharedBodyPartComponent component, PartAddedToBodyEvent args)
+    {
+        var xform = Transform(uid);
+        xform.LocalRotation = 0;
+        xform.AttachParent(args.Body);
```
Comment (metalgearsloth) [inline Content.Shared/Body/Systems/Part/SharedBodyPartSystem.cs:49]: This is the relevant code that will need removing from electro's PR though I haven't sussed out the events that raise it.

File: Content.Shared/Body/Systems/Body/SharedBodySystem.Managerial.cs [hunk 2]
```diff
@@ -0,0 +1,305 @@
...
+        foreach (var hangingPart in GetHangingParts(uid, slot, body))
+        {
+            RemovePart(uid, hangingPart.Key, body);
+        }
+
+        return true;
+    }
+
+    protected bool RemovePartAndRaiseEvents(EntityUid part, BodyPartSlot slot, SharedBodyComponent body)
+    {
```
Comment (mirrorcult) [inline Content.Shared/Body/Systems/Body/SharedBodySystem.Managerial.cs:288]: can this (and the counterpart `AddPartAndRaiseEvents` just get inlined into their respective functions? afaict these are only called in one place internally and it's clearer that way

File: Content.Client/Body/Systems/BodySystem.cs [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+        foreach (var (serverKey, serverSlot) in state.Slots)
+        {
+            var newSlot = new BodyPartSlot(serverSlot);
+            SetSlot(uid, serverKey, newSlot);
+        }
+    }
+
+    // TODO BODY: This is a bit of a hack so the HumanoidAppearanceSystem will still get its events.
+    // In the future the body visuals should probably come from parts directly rather than being "casted" into HumanoidAppearanceSystem
+    private void OnInsertedIntoContainer(EntityUid uid, BodyComponent body, EntInsertedIntoContainerMessage args)
```
Comment (mirrorcult) [inline Content.Client/Body/Systems/BodySystem.cs:42]: wont this double raise events if a part is added using the body functions? same with remove

**Review by mirrorcult — CHANGES_REQUESTED** [review]
mostly just reviewing stuff that I didn't recognize from the old PR. body part slot networking stuff all looks good to me

**Comment by EmoGarbage404** [comment]
@DoctorBeard are you ever coming back to this?

**Pull request closed without merging**